github/install.go

```diff
@@ -0,0 +1,494 @@
+}
+
+const githubApiUrl = "https://api.github.com/"
+
+func fetchMod(slug string) (Mod, error) {
+	var modReleases []ModReleases
+	var mod Mod
+	resp, err := http.Get(githubApiUrl + "repos/" + slug + "/releases")
+	if err != nil {
+		return mod, err
+	}
+
+	if resp.StatusCode == 404 {
+		return mod, fmt.Errorf("mod not found (for URL %v)", githubApiUrl+"repos/"+slug+"/releases")
+	}
+
+	if resp.StatusCode != 200 {
+		return mod, fmt.Errorf("invalid response status %v for URL %v", resp.Status, githubApiUrl+"repos/"+slug+"/releases")
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return mod, err
+	}
+
+	err = json.Unmarshal(body, &modReleases)
+	if err != nil {
+		return mod, err
+	}
+
+	var repoData Repo
+
+	repoResp, err := http.Get(githubApiUrl + "repos/" + slug)
+
+	if err != nil {
+		return mod, err
+	}
+
+	defer repoResp.Body.Close()
+
+	repoBody, err := ioutil.ReadAll(repoResp.Body)
+	if err != nil {
+		return mod, err
+	}
+
+	err = json.Unmarshal(repoBody, &repoData)
+	if err != nil {
+		return mod, err
+	}
+
+	release := modReleases[0]
+
+	mod = Mod{
+		ID:          repoData.Name,
+		Slug:        slug,
+		Team:        repoData.Owner.Login,
+		Title:       repoData.Name,
+		Description: repoData.Description,
+		Published:   repoData.CreatedAt,
+		Updated:     release.CreatedAt,
+		License:     repoData.License,
+		ClientSide:  "unknown",
+		ServerSide:  "unknown",
+		Categories:  repoData.Topics,
+	}
+	if mod.ID == "" {
+		return mod, errors.New("invalid json whilst fetching mod: " + slug)
+	}
+
+	return mod, nil
+
+}
+
+func installMod(mod Mod, pack core.Pack) error {
+	fmt.Printf("Found mod %s: '%s'.\n", mod.Title, mod.Description)
+
+	latestVersion, err := getLatestVersion(mod.Slug, pack, "")
+	if err != nil {
+		return fmt.Errorf("failed to get latest version: %v", err)
+	}
+	if latestVersion.URL == "" {
+		return errors.New("mod is not available for this Minecraft version (use the acceptable-game-versions option to accept more) or mod loader")
+	}
+
+	return installVersion(mod, latestVersion, pack)
+}
+
+func getLatestVersion(slug string, pack core.Pack, branch string) (ModReleases, error) {
+	var modReleases []ModReleases
+	var release ModReleases
+
+	resp, err := http.Get(githubApiUrl + "repos/" + slug + "/releases")
+	if err != nil {
+		return release, err
+	}
+
+	if resp.StatusCode == 404 {
+		return release, fmt.Errorf("mod not found (for URL %v)", githubApiUrl+"repos/"+slug+"/releases")
+	}
+
+	if resp.StatusCode != 200 {
+		return release, fmt.Errorf("invalid response status %v for URL %v", resp.Status, githubApiUrl+"repos/"+slug+"/releases")
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return release, err
+	}
+	err = json.Unmarshal(body, &modReleases)
+	if err != nil {
+		return release, err
+	}
+	for _, r := range modReleases {
+		if r.TargetCommitish == branch {
+			return r, nil
+		}
+	}
+
+	return modReleases[0], nil
```

Comment:
thought about this but if branch = "" it will just return the latest

Reply:
If a branch isn't set it should always just return the first value anyway, no point traversing the list. It is not good UX to just download the latest file when it can't find one on the branch, particularly if it does actually exist but needs to paginate to find it.